keps/sig-apps/3939-include-terminating-pods-as-active/README.md

```diff
@@ -0,0 +1,852 @@
+The Job controller does not typically wait for terminating pods to be marked as failed.  Tensorflow and other ML frameworks may have a requirement that they only want Pods to be started once the other pods are fully terminated.  The following yaml can fit these needs:
+
+This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.
+
+#### Story 2
+
+As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.
+See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.
+
+#### Story 3
+
+As a cloud user, users would want to guarantee that the number of pods that are running includes terminating pods.  In scare compute environments, users may only have a limited amount of nodes and they do not want to try and schedule pods to a new resource.
+Counting terminating pods as active allows for the scheduling of pods to wait until pods are terminated.
+
+See [Option for acknowledging terminating Pods in Deployment rolling update](https://github.com/kubernetes/kubernetes/issues/107920)
+for more examples.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Open Questions on Deployment Controller
+
+The Deployment API is open for discussion.  We put the field in Deployment/ReplicaSet because it is related to RolloutStrategy.
+It is not clear if `recreate` and/or `rollingupdate` need this API for both rollout options.
+
+Another open question is if we want to include Deployments in the initial release of this feature.  There is some discussion about releasing the Job API first and then follow up with Deployment.  
+
+We decided to define the APIs in this KEP as they can utilize the same implementation.
+
+#### Open Questions on Job Controller
+
+With 3329-retriable-and-non-retriable-failures and PodFailurePolicy enabled, terminating pods are only marked as failed once they have been transitioned to failed.  If PodFailurePolicy is disabled, then we mark a terminating pod as failed as soon as deletion is registered.  
+
+Should we add a new field to the status that reflects terminating pods?
+
+[Job controller should wait for Pods to be in a terminal phase before considering them failed or succeeded](https://github.com/kubernetes/kubernetes/issues/116858) is a relevant issue for this case.  
+I am not sure how to handle these two different cases if we want to count terminating pods as active.  
+
+Should we use this feature to help solve 116858?  When this feature toggle is on, then we mark terminating pods only as failed once they are complete regardless of PodFailurePolicy.
```

Comment:
If PodDisruptionConditions is disabled, a pod bound to a no-longer-existing node may be stuck in the Running phase. As a consequence, it will never be replaced, so the whole job will be stuck from making progress.

I guess we could:
1. only enable the new field ActiveUntilTerminal if PodDisruptionConditions is enabled
2. accept there is this potential issue and document
3. make pods as failed in PodGC unconditionally from 1.28, even before graduation of pod failure policy KEP

I would vote for 3 actually, the change feels safe. The fact that PodGC does not mark the pods as failed we could probably call a bug, similarly as done in Kubelet. WDYT @alculquicondor ?

Reply:
We should fix the bug in PodGC (set the phase to Failed), but still tie it to the PodDisruptionConditions feature. Yes, this is a separate issue from this KEP.

Then, we can document that a Pod might be stuck in the Running phase if users enable `TerminatingPodsAsActiveJobs` but disable `PodDisruptionConditions`. This is relevant to the KEP and should go in the "Risks" section.

Reply:
Right, so IIUC no implementation work needed as the Failed phase is already set when PodDisruptionConditions is enabled. We can just update the Risks section, that when PodDisruptionConditions is disabled, orphaned pods might get stuck in the Running phase. As a mitigation we could update the user-facing documentation.

Reply:
Alternatively, we can set phase=Failed when `PodDisruptionConditions` is enabled OR `JobRecreatePodsWhenFailed` is enabled. When `JobRecreatePodsWhenFailed` enabled, but `PodDisruptionConditions`  disabled we would just set the phase, but without adding the condition. 

I think this should be easy to achieve and would allow using these features independently without the significant risk of pods (end entire jobs) getting stuck.  @alculquicondor WDYT?

Reply:
sure, that can work

Reply:
updated the docs.